src/compiler/checker.ts

```diff
@@ -3455,19 +3457,35 @@ namespace ts {
                             }
                             else {
                                 context.approximateLength += 3;
+                                if (!(context.flags & NodeBuilderFlags.NoTruncation)) {
+                                    return createTypeReferenceNode(createIdentifier("..."), /*typeArguments*/ undefined);
+                                }
                                 return createKeywordTypeNode(SyntaxKind.AnyKeyword);
                             }
                         }
                         else {
                             // Since instantiations of the same anonymous type have the same symbol, tracking symbols instead
                             // of types allows us to catch circular references to instantiations of the same anonymous type
-                            if (!context.visitedSymbols) {
-                                context.visitedSymbols = createMap<true>();
+                            if (!context.visitedTypes) {
+                                context.visitedTypes = createMap<true>();
+                            }
+                            if (!context.symbolDepth) {
+                                context.symbolDepth = createMap<number>();
                             }
 
-                            context.visitedSymbols.set(id, true);
+                            const depth = context.symbolDepth.get(id) || 0;
+                            if (depth > 10) {
+                                context.approximateLength += 3;
+                                if (!(context.flags & NodeBuilderFlags.NoTruncation)) {
+                                    return createTypeReferenceNode(createIdentifier("..."), /*typeArguments*/ undefined);

```

Comment:
Hmm this might cause problem with quick Info or something similar where it uses display part kinds where it will coincide with spread operator thus colorization might be odd?

Reply:
We already use `...` in arbitrary places for truncation, and have for around 2 releases? The `displaypartKind` isn't based on the text, but on the function used to print the text, and this is being passed off as an `identifier`, so will be `text`, not `punctuation`.